Nicer IRC notifications

BenWiederhake opened the pull request:
# Why?

## Exhibit A

Too harsh:

> (21:55:35) serenity-ga [~serenity-@40.77.71.119] joined.
> (21:55:35) serenity-ga: tomuta opened PR #4012: failure
> // ...
> (22:18:28) thakis: "opened PR: failure" harsh

## Exhibit B

Too long:

> (22:49:46) serenity-ga [~serenity-@23.101.119.170] joined.
> (22:49:46) serenity-ga: awesomekling pushed master on https://github.com/SerenityOS/serenity/compare/a2cfb7eb9473...81b7c072edb7: success
> (22:49:46) serenity-ga: Subject: AK: Use reference algorithms for LEB128 parsing
> (22:49:46) serenity-ga:  
> (22:49:47) serenity-ga: This fixes a bug in signed LEB128 parsing (sign extension stage)
> (22:49:47) serenity-ga: which would sometimes cause debug info to look very strange. (+ more, maybe)
> (22:49:48) serenity-ga: Details: https://github.com/SerenityOS/serenity/actions/runs/352784189
> (22:49:48) serenity-ga left.

## Exhibit C

Too often:

> (22:56:02) linusg: it's
> (22:56:06) linusg: too
> (22:56:09) linusg: verbose
> (22:57:14) BenW: Really? D:
> (22:57:27) BenW: It's just three lines, the old ones were three lines, too.
> (22:58:04) linusg: but the old ones didn't have "opened" and "synchronized" updates
> (22:58:10) linusg: Just "merged"

# This PR

… introduces a python script that nicely formats the IRC notifications. Examples:

```
(22:27:51) serenity-ga [~serenity-@104.43.251.130] joined.
(22:27:51) serenity-ga: BenWiederhake pushed master: The build passed. https://github.com/SerenityOS/serenity/actions/runs/354757809
(22:27:51) serenity-ga: Subject: TMP: Test it (+3 commits)
(22:27:51) serenity-ga: Details: https://github.com/BenWiederhake/serenity/compare/b59e2d43d3a1...bff54ab3e8dc
(22:27:52) serenity-ga left.
(22:33:27) serenity-ga [~serenity-@40.77.97.227] joined.
(22:33:27) serenity-ga: BenWiederhake opened PR #7: The build passed. https://github.com/SerenityOS/serenity/actions/runs/354763806
(22:33:27) serenity-ga: Subject: Please build me!
(22:33:27) serenity-ga: Details: https://github.com/BenWiederhake/serenity/pull/7
(22:33:28) serenity-ga left.
```

It limits the number of lines to three, never "floods" IRC, and makes everything more grammaticallier correctlierly.

# Why not?

## Python

It adds a new dependency, but this dependency is *only* needed for the Github Action. So I don't know how you feel about that.
This is kind of necessary though, because I really don't want to parse JSON with bash. Finally, writing our own programs in Lagom means that many build errors cause the notification(s) to disappear, so I disregarded that, too. Hence my choice for python: Universal enough, and it's already on the Github Actions image anyway.

## Still too verbose

I didn't change anything about the frequency. However, that could be easily tweaked using the `if:` key. For example, we could only send PR notifications for (un)successful builds. Or none. Or decide in the python script (would need three more lines). I slightly prefer what I implemented (notifications for EVERYTHING), and linus is against it. What do you guys think?

linusg commented:
> grammaticallier correctlierly.

`\o/`

> I slightly prefer what I implemented (notifications for EVERYTHING), and linus is against it

Not necessarily :)
I'm against a high number of individual messages from GA, and as this reduces it to 3 max per event that's perfectly fine.

linusg commented on .github/workflows/cmake.yml:
```diff
@@ -87,29 +87,34 @@ jobs:
       if: github.repository == 'SerenityOS/serenity' && github.ref == 'refs/heads/master' && github.event_name == 'push' && !cancelled()
       with:
-        channel: "#serenityos"
+        channel: "#irc-ga-test"
```
Committing this was probably not intentional ;)

BenWiederhake replied:
Le gasp, my secret test channel has been leaked! Yup, I need to change that.

nico commented:
Ooooh a bikeshed thread! Suggestion:
- drop the first line, it has very few interesting bits (or very few bits that are so interesting that they need to be surfaced on irc)
- combine lines "Subject: foo" and "Details: bar" into a single line "foo: bar"

awesomekling commented:
I have no issue with using Python here. The sooner this goes in, the better. The current IRC notifications are way too much.

alimpfard commented:
If I may join the bikeshedding,

One more request: Please skip the entire thing if the last commit contains `skip ci` (just like travis did).
\*cough cough\* maybe don't notify on PRs (or just make them single-line?)

BenWiederhake commented:
Changes since the last push:
- Use `#serenityos` instead of my super-secret test channel (hi \0!)
- Significantly shorter notifications, even shorter than those three lines: ONE line!
- Ignore pushes(!) that end in a commit that contains "skip ci" or "ci skip" somewhere. I'm not sure what you need that for, but okay, why not!

Here's an example of how it *should* look like:

On a push:
```
serenity-ga [~serenity-@104.43.251.130] joined.
serenity-ga: Meta: Nicer IRC notifications (+20 commits) (pushed master: The build passed.) https://github.com/BenWiederhake/serenity/compare/6812119f01b0...cae7f7f39af9 https://github.com/SerenityOS/serenity/actions/runs/354757809
serenity-ga left.
```

On a PR:
```
serenity-ga [~serenity-@104.43.251.130] joined.
serenity-ga: Meta: Improve all the things (BenWiederhake created: The build passed.) https://github.com/BenWiederhake/serenity/pull/7 https://github.com/SerenityOS/serenity/actions/runs/354757809
serenity-ga left.
```

alimpfard commented:
>  I'm not sure what you need that for, but okay, why not!

I tend to use branches themselves to keep uncommitted stuff, much less confusing than the mess that `git stash` is, and I don't like the ci wasting time building and testing something I don't even consider working.
> `alias gwip='git add -A; git rm $(git ls-files --deleted) 2>/dev/null; gcm "--wip-- [skip ci]"'`

nico commented:
@alimpfard I think the CI only builds stuff in a PR though, no? If you just have a bunch of branches in your repo, would they be built too?

alimpfard commented:
Currently, everything pushed is built.
You even get email notifications for them

BenWiederhake commented:
> Currently, everything pushed is built.
> You even get email notifications for them

Oh, *that's* what you intended! Well, right now it only disables the IRC notifications, and the check for that is run at the very end of the build.

You *could* in theory disable Github Actions entirely, but that seems like a bad idea.

One could add a step in the beginning (that needs python or `jq`) that can stop the entire build – however, you would still get an e-mail.

In your [account settings](https://github.com/settings/notifications) you can disable those e-mails, or only be notified of broken builds.

The previous two together would mean that "[skip ci]" pushes take only a few seconds to run, count as a success, and you don't get an e-mail. I think that's the best achievable.
